engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt

```diff
@@ -0,0 +1,185 @@
+    fun setUp() {
+        mockWebServer.start()
+    }
+
+    @After
+    fun tearDown() {
+        mockWebServer.shutdown()
+    }
+
+    @Test
+    fun `Converts a success response with null body into an ShiftsResult exception`() {
+        val response: Response<List<Shift>> = Response.success(null)
+        val shiftsResult = response.toShiftsResult()
+        assertThat(shiftsResult).isEqualTo(ShiftsResult.Exception.MissingResponseSuccessBody)
+    }
+
+    @Test
+    fun `Converts a success response with empty body into an ShiftsResult success with an empty list`() {
+        val shiftsResult = Response.success(emptyList<Shift>()).toShiftsResult()
+        assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(emptyList()))
+    }
+
+    @Test
+    fun `Converts a success response with one item shift into an ShiftsResult success with an one item shifts list`() {
+        val shiftsResult = Response.success(ONE_ITEM_SHIFTS).toShiftsResult()
+        assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(ONE_ITEM_SHIFTS))
+    }
+
+    @Test
+    fun `awaitShiftsResult returns exception when call responds with HTTP 200 and invalid JSON`() {
+        withResponse(httpStatusCode = /* not relevant here */ HttpURLConnection.HTTP_OK, shiftsJson = INVALID_ONE_ITEM_SHIFTS_JSON) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isInstanceOf(ShiftsResult.Exception::class.java)
+            assertThat((shiftsResult as ShiftsResult.Exception).throwable).isInstanceOf(JsonDataException::class.java)
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns exception when shifts JSON is empty`() {
+        withResponse(httpStatusCode = /* not relevant here */ HttpURLConnection.HTTP_OK, shiftsJson = EMPTY_STRING) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isInstanceOf(ShiftsResult.Exception::class.java)
+            assertThat((shiftsResult as ShiftsResult.Exception).throwable).isInstanceOf(EOFException::class.java)
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns success and empty list when call responds with HTTP 200 and empty array JSON`() {
+        withResponse(HttpURLConnection.HTTP_OK, EMPTY_ARRAY_SHIFTS_JSON) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(emptyList()))
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns success when call responds with HTTP 200`() {
+        withResponse(HttpURLConnection.HTTP_OK, VALID_ONE_ITEM_SHIFTS_JSON) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Success(ONE_ITEM_SHIFTS))
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns error when call responds with HTTP 300`() {
+        withResponse(HttpURLConnection.HTTP_MULT_CHOICE, VALID_ONE_ITEM_SHIFTS_JSON) { call ->
```

Comment:
`VALID_ONE_ITEM_SHIFTS_JSON` doesn't seem relevant for the error cases. It's also unlikely the error bodies will contain a valid JSON response in practice. In this case a default value for the `shiftsJson` parameter would be good.

Reply:
:heavy_check_mark: Done in d616cee9a6bd0f59e474a9a4e7c92efc45792b7a.